source/fe/mapping_q_cache.cc

```diff
@@ -125,6 +184,319 @@ MappingQCache<dim, spacedim>::initialize(
+
+              fe_values = std::make_unique<FEValues<dim, spacedim>>(
+                mapping, fe_temp, quadrature, update_quadrature_points);
+            }
+
+          fe_values->reinit(cell);
+        }
+
+      std::vector<Point<spacedim>> result;
+
+      // Step 2b) read of quadrature points in the relative displacement case
+      if (vector_describes_relative_displacement || (is_relevant_cell == false))
+        {
+          if (mapping_q_generic != nullptr &&
+              this->get_degree() == mapping_q_generic->get_degree())
+            result =
+              mapping_q_generic->compute_mapping_support_points(cell_tria);
+          else
+            result = fe_values_all.get()->get_quadrature_points();
+
+          if (is_relevant_cell == false)
+            return result;
+        }
+      else
+        {
+          result.resize(
+            Utilities::pow<unsigned int>(this->get_degree() + 1, dim));
+        }
+
+      // Step 2c) read global vector and adjust points accordingly
+      if ((is_fe_q || is_fe_dgq) && fe.degree == this->get_degree())
+        {
+          AssertDimension(dim, spacedim); // TODO: codim-1
+
+          // case 1: FE_Q or FE_DGQ with same degree as this class has; this
+          // is the simple case since no interpolation is needed
+          std::vector<types::global_dof_index> dof_indices(
+            fe.n_dofs_per_cell());
+          cell->get_dof_indices(dof_indices);
+
+          for (unsigned int i = 0; i < dof_indices.size(); ++i)
+            {
+              const auto id = fe.system_to_component_index(i);
+
+              if (is_fe_q)
+                {
+                  // case 1a: FE_Q
+                  if (vector_describes_relative_displacement)
+                    result[id.second][id.first] +=
+                      vector_ghosted(dof_indices[i]);
+                  else
+                    result[id.second][id.first] =
+                      vector_ghosted(dof_indices[i]);
+                }
+              else
+                {
+                  // case 1b: FE_DGQ
+                  if (vector_describes_relative_displacement)
+                    result[lexicographic_to_hierarchic_numbering[id.second]]
+                          [id.first] += vector_ghosted(dof_indices[i]);
+                  else
+                    result[lexicographic_to_hierarchic_numbering[id.second]]
+                          [id.first] = vector_ghosted(dof_indices[i]);
+                }
+            }
+        }
+      else
+        {
+          Assert(false, ExcNotImplemented()); // TODO: general case
+        }
+
+      return result;
+    });
+}
+
+
+
+template <int dim, int spacedim>
+template <typename VectorType>
+void
+MappingQCache<dim, spacedim>::initialize(
+  const Mapping<dim, spacedim> &   mapping,
+  const DoFHandler<dim, spacedim> &dof_handler,
+  const MGLevelObject<VectorType> &vector,
+  const bool                       vector_describes_relative_displacement)
```

Comment:
Since this is more or less the same as the above function but with multiple vectors being accessed, I tackle this once the above function is ready.